### PR TITLE
WIP: Push activation

### PR DESCRIPTION
**Pull request by tcard**
Fixes https://github.com/ably/ably-js/issues/429.

### CODE REVIEW

File: browser/lib/util/ulid.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		}
+		if (allowInsecure) {
+			try {
+				console.error("secure crypto unusable, falling back to insecure Math.random()!");
+			} catch (e) {}
+			return function () {
+				return Math.random();
+			};
+		}
+		throw createError("secure crypto unusable, insecure Math.random not allowedW");
```
Comment (mattheworiordan) [inline browser/lib/util/ulid.js]: Typo, trailing `W`

Comment (tcard) [reply browser/lib/util/ulid.js]: Well, that makes for a juicy PR upstream :)

https://github.com/ulid/javascript/pull/59

**Comment by paddybyers** [comment]
Looking good so far.

I don't like the fact that we don't have the state machine in node. I know there's no transport currently that we can use for this, but I'd prefer that the state machine implementation was just a generic part of `Push`, and it calls the initialiser of some underlying object/transport when activated, which throws an exception in node to say it's not supported.

In a similar vein, the actual code implementing the subscription (https://github.com/ably/ably-js/blob/dc984ae15e3d993986127f1e6f9416c9e6448823/browser/lib/client/push.js#L132-L168) is embedded within the state machine, whereas I'd prefer that all of these platform-specific primitive operations were encapsulated in a `PushTransport` class or something, so the state machine could be used with some other transport. For example we could, for testing purposes, implement a transport that uses the `ablyChannel` mechanism so we can exercise the state machine in node tests. Or, is it possible for us to have a fcm client in node?

Also in js can we use `const` where possible pls.

**Comment by SimonWoolf** [comment]
> Also in js can we use const where possible pls.

ably-js nominally still supports back to IE8 and node 0.12, we're a long way from being able to use `const`

**Comment by paddybyers** [comment]
> ably-js nominally still supports back to IE8 and node 0.12, we're a long way from being able to use const

Oops, yes. In that case there's lots of `let` and `const` in `Push.js` that needs fixing.

**Comment by tcard** [comment]
Sorry, fixed those.

Do we have some kind of syntax checker in place?

**Comment by paddybyers** [comment]
> Do we have some kind of syntax checker in place?

I don't see any `jshint` task or equivalent in the gruntfile. @SimonWoolf ?

**Comment by SimonWoolf** [comment]
nope, as far as I know we've never had a linter beyond the syntax checking provided by the closure-compiler minifier. I agree that it would be a good idea.

**Comment by funkyboy** [comment]
Up to this point push tests pass both on `nodeunit` and `karma`. 
Tests have been run on Chrome 63.0.3239.132 and Firefox 57.0.4 (both on macOS 10.12.6).
Tests are still heterogeneous tho, some in `JSON` some in `msgpack`.

**Comment by QuintinWillison** [comment]
Replaced by #648

**Comment by jessica02200** [comment]
How does work?

**Pull request closed without merging**